File: intermediate/final/prog/stretch_c3.py
import os
import sys

def toRowCol(r, c, s):
	ri = (s-1) // c
	ci = (s-1) % c
	return [ri, ci]

def toSeqNum(r, c, ri, ci):
	return ri * c + ci + 1

def isValid(r, c, ri, ci):
	return 0 <= ri and ri < r and 0 <= ci and ci < c

def nextPiece(curr):
	ps = ['A','B','C']
	if not curr:
		return 'A'
	return ps[(ord(curr) - ord('A') + 1) % 3]

def isEnd(r, c, s):
	rc = toRowCol(r, c, s)
	return rc[1] == c - 1

def endPole(r, c, begPole, piece):
	rc = toRowCol(r, c, begPole)
	rl = rc[0]
	cl = rc[1]
	rr = 0
	cr = 0
	if piece == 'A':
		rr = rl
		cr = cl + 2
	elif piece == 'B':
		rr = rl + 1
		cr = cl + 1
	elif piece == 'C':
		rr = rl + 2
		cr = cl + 1
	return toSeqNum(r, c, rr, cr) if isValid(r, c, rr, cr) else 0

def canPlace(r, c, blocks, begPole, piece):
	rc = toRowCol(r, c, begPole)
	ri = rc[0]
	ci = rc[1]
	if piece == 'A':
		if not isValid(r,c,ri,ci) or blocks[toSeqNum(r,c,ri,ci)] or not isValid(r,c,ri,ci+1) or blocks[toSeqNum(r,c,ri,ci+1)] or not isValid(r,c,ri,ci+2) or blocks[toSeqNum(r,c,ri,ci+2)]:
			return False
	elif piece == 'B':
		if not isValid(r,c,ri,ci) or blocks[toSeqNum(r,c,ri,ci)] or not isValid(r,c,ri+1,ci) or blocks[toSeqNum(r,c,ri+1,ci)] or not isValid(r,c,ri+1,ci+1) or blocks[toSeqNum(r,c,ri+1,ci+1)]:
			return False
	elif piece == 'C':
		if not isValid(r,c,ri,ci) or blocks[toSeqNum(r,c,ri,ci)] or not isValid(r,c,ri,ci+1) or blocks[toSeqNum(r,c,ri,ci+1)] or not isValid(r,c,ri+1,ci+1) or blocks[toSeqNum(r,c,ri+1,ci+1)] or not isValid(r,c,ri+2,ci+1) or blocks[toSeqNum(r,c,ri+2,ci+1)]:
			return False
	return True

if __name__=="__main__":
	N = 5 if len(sys.argv) < 2 else int(sys.argv[1])
	for i in range(N):
		# Process input data
		nums = [int(x) for x in input().split()]
		r = nums[0]
		c = nums[1]
		s = nums[2]
		n = nums[3]
		blocks = [False]*(r*c+1)
		for j in range(n):
			blocks[nums[j+4]] = True
		# Complete Search - whenever a piece can be placed, move on to the next step
		piece = None
		res = ''
		ep = s
		found = False
		while not found:
			for i in range(3):
				if found:
					break
				piece = nextPiece(piece)
				if canPlace(r, c, blocks, ep, piece):
					res += piece
					ep = endPole(r, c, ep, piece)
					if isEnd(r, c, ep):
						found = True
					else:
						ep += 1
					break
		print(res)

File: intermediate/final/prog/stretch_c3.cpp
#include <iostream>
#include <vector>
using namespace std;

pair<int,int> toRowCol(int r, int c, int s) {
	int ri = (s-1) / c;
	int ci = (s-1) % c;
	return {ri, ci};
}

int toSeqNum(int r, int c, int ri, int ci) {
	return ri * c + ci + 1;
}

bool isValid(int r, int c, int ri, int ci) {
	return 0 <= ri && ri < r && 0 <= ci && ci < c;
}

char nextPiece(char curr) {
	if(!curr)
		return 'A';
	return (curr - 'A' + 1) % 3 + 'A';
}

bool isEnd(int r, int c, int s) {
	auto rc = toRowCol(r, c, s);
	return rc.second == c - 1;
}

int endPole(int r, int c, int begPole, char piece) {
	auto rc = toRowCol(r, c, begPole); int rl = rc.first; int cl = rc.second;
	int rr, cr;
	switch(piece) {
	case 'A': rr = rl;   cr = cl + 2; break;
	case 'B': rr = rl+1; cr = cl + 1; break;
	case 'C': rr = rl+2; cr = cl + 1; break;
	}
	return isValid(r, c, rr, cr) ? toSeqNum(r, c, rr, cr) : 0;
}

bool canPlace(int r, int c, vector<bool>& blocks, int begPole, char piece) {
	auto rc = toRowCol(r, c, begPole);
	int ri = rc.first;
	int ci = rc.second;
	switch(piece) {
	case 'A':
		if( !isValid(r, c, ri, ci  ) || blocks[toSeqNum(r, c, ri, ci  )] || // left-pole (center cell)
			!isValid(r, c, ri, ci+1) || blocks[toSeqNum(r, c, ri, ci+1)] || // ‘A’
			!isValid(r, c, ri, ci+2) || blocks[toSeqNum(r, c, ri, ci+2)] )  // right-pole
			return false;
		break;
	case 'B':
		if(	!isValid(r, c, ri  , ci  )||blocks[toSeqNum(r, c, ri  , ci  )] || // left-pole (center cell)
			!isValid(r, c, ri+1, ci  )||blocks[toSeqNum(r, c, ri+1, ci  )] || // ‘B’
			!isValid(r, c, ri+1, ci+1)||blocks[toSeqNum(r, c, ri+1, ci+1)] )  // right-pole
			return false;
		break;
	case 'C':
		if( !isValid(r, c, ri  , ci  )||blocks[toSeqNum(r, c, ri  , ci  )] || // left-pole (center cell)
		    !isValid(r, c, ri  , ci+1)||blocks[toSeqNum(r, c, ri  , ci+1)] || // Upper ‘C’
		    !isValid(r, c, ri+1, ci+1)||blocks[toSeqNum(r, c, ri+1, ci+1)] || // Lower ‘C’
		    !isValid(r, c, ri+2, ci+1)||blocks[toSeqNum(r, c, ri+2, ci+1)] )  // right-pole
			return false;
		break;
	}
	return true;
}

int main(int argc, char* argv[]) {
	int N = argc > 1 ? atoi(argv[1]) : 5;
	for(int i = 0; i < N; ++i) {
		// Process input data
		int r,c,s,n;
		cin >> r >> c >> s >> n;
		vector<bool> blocks(r*c+1, false);
		for(int j = 0; j < n; ++j) {
			int b;
			cin >> b;
			blocks[b] = true;
		}
		// Complete Search - whenever a piece can be placed, move on to the next step
		char piece = '\0';
		string res;
		int ep = s;
		bool found = false;
		while(!found) {
			for(int i = 0; i < 3 && !found; ++i) {
				piece = nextPiece(piece);
				if(canPlace(r, c, blocks, ep, piece)) {
					res.push_back(piece);
					ep = endPole(r, c, ep, piece);
					if(isEnd(r, c, ep))
						found = true;
					else
						++ep;
					break;
				}
			}
		}
		cout << res << endl;
	}
	return 0;
}
